lib_eio/sem_state.ml

```diff
@@ -0,0 +1,181 @@
+       but seeing we did that first it will handle the wake-up. *)
+    ()
+
+(* [true] on success, or [false] if we need to suspend.
+   You MUST call [suspend] iff this returns [false].
+   The reason for splitting this is because e.g. [Semaphore] needs to get
+   the continuation for the fiber between [acquire] and [suspend]. *)
+let acquire t =
+  let s = Atomic.fetch_and_add t.state (-1) in
+  (* We got a resource if we decremented *to* a non-negative number,
+     which happens if we decremented *from* a positive one. *)
+  s > 0
+
+let suspend t k : request option =
+  let (segment, cell) = Cells.next_suspend t.cells in
+  if Atomic.compare_and_set cell Empty (Request k) then Some (t, segment, cell)
+  else match Atomic.get cell with
+    | Finished ->
+      (* We got resumed before we could add the waiter. *)
+      k ();
+      None
+    | Cancelling | Request _ | Empty ->
+      (* These are unreachable from the previously-observed non-Empty state
+         without us taking some action first *)
+      assert false
+
+let release t =
+  let s = Atomic.fetch_and_add t.state (+1) in
+  if s < 0 then (
+    (* We incremented from a negative value.
+       We are therefore responsible for waking one waiter. *)
+    resume t
+  )
+
+let cancel (t, segment, cell) =
+  match (Atomic.get cell : cell) with
+  | Request _ as old ->
+    if Atomic.compare_and_set cell old Cancelling then (
+      (* Undo the effect of [acquire] by incrementing the counter.
+         As always, if we increment from a negative value then we need to resume one waiter. *)
+      let need_resume = Atomic.fetch_and_add t.state (+1) < 0 in
+      if need_resume then (
+        if Atomic.compare_and_set cell Cancelling Finished then (
+          (* The normal case. We resumed ourself by cancelling. *)
+          Cells.cancel_cell segment
+        ) else (
+          (* [release] got called at the same time and it also needed to resume one waiter.
+             So we call [resume] to handle the extra one, in addition to resuming ourself. *)
+          resume t
+        )
+      ) else (
+        (* This can only happen if [release] ran at the same time and incremented the counter
+           before we did. Since we were suspended, and later we saw the counter
+           show that no one was, it must have decided to wake us. Either it has placed Finished
+           in the cell, or it's about to do so. Either way, we discharge the obligation to
+           wake someone by resuming ourself with a cancellation. *)
+      );
+      true
+    ) else false          (* We got resumed first *)
+  | Finished -> false     (* We got resumed first *)
+  | Cancelling -> invalid_arg "Already cancelling!"
+  | Empty ->
+    (* To call [cancel] the user needs a [request] value,
+       which they only get once we've reached the [Request] state.
+       [Empty] is unreachable from [Request]. *)
+    assert false
```

Comment:
Hmm... Speaking of reducing states, I wonder whether `Empty` and `Cancelling` states could actually be merged.

Is this the only place where they are handled differently?  And both indicate a programming error?

Reply:
Good point - done :-)